firedrake/slate/slac/compiler.py

```diff
@@ -224,7 +229,24 @@ def generate_kernel_ast(builder, statements, declared_temps):
     # Now we wrap up the kernel ast as a PyOP2 kernel and include the
     # Eigen header files
     include_dirs = builder.include_dirs
-    include_dirs.extend(["%s/include/eigen3/" % d for d in PETSC_DIR])
+
+    def extract_eigen_include_dir(PETSC_DIR, PETSC_ARCH):
+        try:
+            with open(os.path.join(PETSC_ARCH, "lib", "petsc", "conf", "petscvariables") if PETSC_ARCH
+                      else os.path.join(PETSC_DIR, "lib", "petsc", "conf", "petscvariables")) as file:
+                for line in file:
+                    if line.find("EIGEN_INCLUDE") == 0:
+                        include_dir = line[18:].rstrip()
+                        return include_dir
+        except FileNotFoundError:
+            raise FileNotFoundError("Could not find %s/lib/petsc/conf/petscvariables." % PETSC_ARCH if PETSC_ARCH else PETSC_DIR)
+
+        file.close()
+        raise ValueError(""" Could not find Eigen configuration in %s/lib/petsc/conf/petscvariables.
+                             Check PETSc was built with the correct options.""" % PETSC_ARCH)
+
+    include_dir = extract_eigen_include_dir(PETSC_DIR, PETSC_ARCH)
```

Comment:
Please only do this once (so at module level) rather than once per kernel invocation: I really don't want to parse a file on disk every time. To avoid hitting filesystems too hard on supercomputers please also do it as:
```python
if COMM_WORLD.rank == 0:
   ...
   COMM_WORLD.bcast(...)
else:
   ...
```

Reply:
So there must be things I'm still missing about how Firedrake works.
Isn't the compiler only called at most once in a run ? or even not if the kernel is in the cache ?
In master, the kernel explicitly includes $PETSC_DIR/includes...Eigen/Dense.h. How is that better in terms of filesystem access ?

Reply:
Once per run per kernel.

Reply:
Presently this hits the file system additionally to determine the same (include) information once every time you call generate_kernel_ast. I would like not to do that.

Reply:
So I do it on one proc only now. I have left it in the concerned function though - do you (really) want to add a global variable containing this information that is used nowhere else ?

Reply:
Yes. Because hitting the file system unnecessarily on a supercomputer can take 30minutes.

Reply:
Ok. I've just realised the function was actually called several times. Made the change.